### PR TITLE
Add /v3/stacks endpoint; Fixes #1172

### LINKED ISSUE #1172: [Feature]: cf user can list supported stacks
### Blockers/Dependencies

_No response_

### Background

**As a** cf user
**I want** to list the supported stacks
**So that** I understand how my droplet will be built

### Acceptance Criteria

**GIVEN** a korifi cluster
**WHEN I** run `cf stacks`
**THEN I** see a list of supported stacks with all fields filled out (right now this would be only `io.buildpacks.stacks.bionic`)

### Dev Notes

Here is the v3 documentation for the endpoint: https://v3-apidocs.cloudfoundry.org/version/3.117.0/index.html#list-stacks. **Don't worry about implementing pagination, order_by or any filtering**. You can assume no parameters

Tasks:
- [ ] Add a handler for GET /v3/stacks
- [ ] Add a repository for Stacks that translates `BuildReconcilerInfo.status.stacks` into `StackRecord` objects
- [ ] Add a presenter for `StackRecord` and use it in the handler

### PR DESCRIPTION
**Pull request by benjaminguttmann-avtq**
## Is there a related GitHub Issue?

#1172

## What is this change about?
Adding endpoint /v3/stacks

## Does this PR introduce a breaking change?
No

## Things to remember

Code is inspired by the buildpacks implementation.
As the issue states to only focus on providing the endpoint without any filters, I will work on implementing the filters after this one is merged

### CODE REVIEW

File: api/repositories/stack_repository.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, apierrors.NewResourceNotReadyError(fmt.Errorf("BuilderInfo %q not found in namespace %q", r.builderName, r.rootNamespace))
+		}
+
+		return nil, apierrors.FromK8sError(err, StackResourceType)
+	}
+
+	if !meta.IsStatusConditionTrue(builderInfo.Status.Conditions, korifiv1alpha1.StatusConditionReady) {
+		var conditionNotReadyMessage string
```
Comment (danail-branekov) [inline api/repositories/stack_repository.go:77]: I would suggest skipping this complex not-ready check. If the builderInfo is not ready, then the error could just say `BuilderInfo %q not ready` without additional details. They do not make sense to cf api users and they would have to contact the operator anyway. kubernetes users can just peak into the builder info status and figure it out themselves

**Review by danail-branekov — CHANGES_REQUESTED** [review]

**Comment by danail-branekov** [comment]
Looks good overall, put some minor comments

File: api/handlers/stack.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		serverURL: serverURL,
+		stackRepo: stackRepo,
+	}
+}
+
+func (h *Stack) get(r *http.Request) (*routing.Response, error) {
+	authInfo, _ := authorization.InfoFromContext(r.Context())
+	logger := logr.FromContextOrDiscard(r.Context()).WithName("handlers.build.list")
+
+	stacks, err := h.stackRepo.GetStacks(r.Context(), authInfo)
```
[the remark this answers is not in the document]
Comment (benjaminguttmann-avtq) [reply api/handlers/stack.go]: Renamed GetStacks To ListStacks accordingly

File: api/handlers/stack_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+		apiHandler := NewStack(*serverURL, stackRepo)
+		routerBuilder.LoadRoutes(apiHandler)
+	})
+
+	JustBeforeEach(func() {
+		routerBuilder.Build().ServeHTTP(rr, req)
+	})
+
+	Describe("the GET /v3/stacks endpoint", func() {
```
[the remark this answers is not in the document]
Comment (benjaminguttmann-avtq) [reply api/handlers/stack_test.go:35]: added the requested When

File: api/presenter/stack.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	toReturn := StackResponse{
+		GUID:      stackRecord.GUID,
+		CreatedAt: formatTimestamp(&stackRecord.CreatedAt),
+		UpdatedAt: formatTimestamp(stackRecord.UpdatedAt),
+		Name:      stackRecord.Name,
+		Metadata: Metadata{
+			Labels:      map[string]string{},
+			Annotations: map[string]string{},
+		},
+		Links: map[string]Link{},
```
[the remark this answers is not in the document]
Comment (benjaminguttmann-avtq) [reply api/presenter/stack.go]: added the self.href

File: api/repositories/stack_repository.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	rootNamespace     string
+}
+
+type StackRecord struct {
+	GUID        string
+	CreatedAt   time.Time
+	UpdatedAt   *time.Time
+	Name        string
+	Description string
+	Labels      map[string]string
```
[the remark this answers is not in the document]
Comment (benjaminguttmann-avtq) [reply api/repositories/stack_repository.go]: removed labels and annotations

File: api/repositories/stack_repository.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+		ctx,
+		types.NamespacedName{
+			Namespace: r.rootNamespace,
+			Name:      r.builderName,
+		},
+		&builderInfo,
+	)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil, apierrors.NewResourceNotReadyError(fmt.Errorf("BuilderInfo %q not found in namespace %q", r.builderName, r.rootNamespace))
```
[the remark this answers is not in the document]
Comment (benjaminguttmann-avtq) [reply api/repositories/stack_repository.go]: removed the if and returned the `apierrors.FromK8sError(err, StackResourceType)`

File: api/repositories/stack_repository.go (hunk 1 above)
Comment (benjaminguttmann-avtq) [reply api/repositories/stack_repository.go]: Adjusted the error message and removed the complex not-ready check

File: api/repositories/stack_repository.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+	if !meta.IsStatusConditionTrue(builderInfo.Status.Conditions, korifiv1alpha1.StatusConditionReady) {
+		return nil, apierrors.NewResourceNotReadyError(fmt.Errorf("BuilderInfo %q not ready", r.builderName))
+	}
+
+	return builderInfoToStackRecords(builderInfo), nil
+}
+
+func builderInfoToStackRecords(info korifiv1alpha1.BuilderInfo) []StackRecord {
+	return slices.Collect(it.Right(it.Map2(slices.All(info.Status.Stacks), func(i int, s korifiv1alpha1.BuilderInfoStatusStack) (int, StackRecord) {
```
[the remark this answers is not in the document]
Comment (benjaminguttmann-avtq) [reply api/repositories/stack_repository.go]: adjusted accordingly

**Review by danail-branekov — APPROVED** [review]

**Comment by benjaminguttmann-avtq** [comment]
@danail-branekov @georgethebeatle can someone let me know what failed, I do not have access to check it myself and locally all tests work including e2e tests

**Comment by georgethebeatle** [comment]
@benjaminguttmann-avtq our pipeline is going through some turbulence right now, not your fault. In the meantime we would appreciate it if you squashed your commits into one :)

**Pull request merged**